### PR TITLE
Fix for not being able to save decimal value

### PR DESCRIPTION
**Pull request by RobertCopilau**
http://issues.umbraco.org/issue/UAASSCRUM-1401

### CODE REVIEW

File: src/Umbraco.Core/ObjectExtensions.cs [hunk 1]
```diff
@@ -212,7 +212,7 @@ public static Attempt<object> TryConvertTo(this object input, Type target)
                 var convertible2 = input as IConvertible;
                 if (convertible2 != null)
                 {
-                    return Attempt.Succeed(Convert.ChangeType(convertible2, target));
+                    return Attempt.Succeed(Convert.ChangeType(convertible2, Nullable.GetUnderlyingType(target) ?? target));
```
Comment (zpqrtbnk) [inline src/Umbraco.Core/ObjectExtensions.cs]: I assume that `Nullable.GetUnderlyingType()` is considered expensive, as we have method `GetCachedGenericNullableType()` which is used above and caches the underlying type. Might be good to reuse it here too.

Comment (RobertCopilau) [reply src/Umbraco.Core/ObjectExtensions.cs]: Hmm, alright - I'll see if we can use that instead.

File: src/Umbraco.Core/ObjectExtensions.cs [hunk 2]
```diff
@@ -156,6 +156,11 @@ public static Attempt<object> TryConvertTo(this object input, Type target)
                         if (inner.Success)
                         {
                             input = inner.Result; // Now fall on through...
+
```
Comment (zpqrtbnk) [inline src/Umbraco.Core/ObjectExtensions.cs]: Ah we can't really be that specific. Going to have the same issue with ints, anything really. What happens is... if `target` is nullable, then we already convert `input` to the nullable underlying type, and change `input` with that value. So... right before the line near the end that does `var convertible2 = ...`, you need to handle the special case, with something along

```
if (target.IsGenericType && GetCachedGenericNullableType(target) != null)
{
  // cannot Convert.ChangeType as that does not work with nullable
  // input has already been converted to the underlying type - just
  // return input, there's an implicit conversion from T to T? anyways
  return Attempt.Succeed(input);
}
```

so basically the only thing we change is ... if `target` is a nullable there's no need to go through the `IConvertible` thing, just return `input` should be OK.

Making sense?

Comment (RobertCopilau) [reply src/Umbraco.Core/ObjectExtensions.cs]: Owww, yes, much simple and elegant. Will take care of it first thing in the morning.

Comment (Shazwazza) [reply src/Umbraco.Core/ObjectExtensions.cs]: I've made this changed and pushed it. I've moved all the TryConvertTo tests to the ObjectExtensionsTests where the others live.

All tests pass now, but I've also added a TODO note if @zpqrtbnk can have a look? We do a check for input of `string` and then check `IsNullOrEmpty` and then specifically check for `datetime` and `bool` and return null but that seems odd right? Why the bool/datetime check, if the string IsNullOrEmpty and the destintation isn't a string, but its nullable, wouldn't it always just be null ?

File: src/Umbraco.Core/ObjectExtensions.cs (hunk 1 above)
Comment (Shazwazza) [reply src/Umbraco.Core/ObjectExtensions.cs]: With the other change made, I don't think this line is required. I've changed it back to what it was before and all tests pass. I think with the new check above `if (target.IsGenericType && GetCachedGenericNullableType(target) != null)` this is not needed?

File: src/Umbraco.Tests/ObjectExtensionsTests.cs [hunk 3]
```diff
@@ -164,28 +168,166 @@ public virtual void CanConvertObjectToSameObject()
...
+            Assert.AreEqual(new DateTime(2016, 6, 7), conv.Result);
+        }
+
+        [Test]
+        public void ConvertToNullableDateTimeTest()
+        {
+            var conv = "2016-06-07".TryConvertTo<DateTime>();
+            Assert.IsTrue(conv);
+            Assert.AreEqual(new DateTime(2016, 6, 7), conv.Result);
+        }
```
Comment (bjarnef) [inline src/Umbraco.Tests/ObjectExtensionsTests.cs]: Shouldn't these two test methods of TryConvertTo `DateTime`  / `DateTime?` be opposite?

Comment (nul800sebastiaan) [reply src/Umbraco.Tests/ObjectExtensionsTests.cs]: Sure, the naming is a bit off, but at least both cases are covered. :-) I'll update.

**Pull request merged**